File: problems/0278.0_First_Bad_Version.py
'''
Binary Search
T: O(logN)
S: O(1)

Runtime: 36 ms, faster than 32.83% of Python3 online submissions for First Bad Version.
Memory Usage: 14.3 MB, less than 42.48% of Python3 online submissions for First Bad Version.
'''
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return an integer
# def isBadVersion(version):

class Solution:
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        l, r = 1, n
        while l <= r:
            mid = (l + r) // 2
            if isBadVersion(mid):
                r = mid - 1
            else:
                l = mid + 1
        return r + 1
        

File: cpp/0278.0_First_Bad_Version.cpp
/*
执行用时：0 ms, 在所有 C++ 提交中击败了100.00% 的用户
内存消耗：5.8 MB, 在所有 C++ 提交中击败了46.32% 的用户
通过测试用例：22 / 22
*/
// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        int l = 1, r = n;
        while (l < r) {
            int mid = (r - l ) / 2 + l;
            if (isBadVersion(mid)) {
                r = mid;
            } else {
                l = mid + 1;
            }
        }
        return r;
    }
};

